lib/Dialect/HW/Transforms/HWAggregateToComb.cpp

```diff
@@ -0,0 +1,186 @@
+
+namespace circt {
+namespace hw {
+#define GEN_PASS_DEF_HWAGGREGATETOCOMB
+#include "circt/Dialect/HW/Passes.h.inc"
+} // namespace hw
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+
+namespace {
+template <typename OpTy>
+struct HWArrayCreateLikeOpConversion : OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  using OpAdaptor = typename OpConversionPattern<OpTy>::OpAdaptor;
+  LogicalResult
+  matchAndRewrite(OpTy op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Lower to concat.
+    auto inputs = adaptor.getInputs();
+    SmallVector<Value> results;
+    for (auto input : inputs)
+      results.push_back(rewriter.getRemappedValue(input));
+    rewriter.replaceOpWithNewOp<comb::ConcatOp>(op, results);
+    return success();
+  }
+};
+
+struct HWAggregateConstantOpConversion
+    : OpConversionPattern<hw::AggregateConstantOp> {
+  using OpConversionPattern<hw::AggregateConstantOp>::OpConversionPattern;
+
+  static LogicalResult peelAttribute(Location loc, Attribute attr,
+                                     ConversionPatternRewriter &rewriter,
+                                     SmallVector<Value> &results) {
+    SmallVector<Attribute> worklist;
+    worklist.push_back(attr);
+
+    while (!worklist.empty()) {
+      auto current = worklist.pop_back_val();
+      if (auto innerArray = dyn_cast<ArrayAttr>(current)) {
+        for (auto elem : llvm::reverse(innerArray))
+          worklist.push_back(elem);
+        continue;
+      }
+
+      if (auto intAttr = dyn_cast<IntegerAttr>(current)) {
+        results.push_back(rewriter.create<hw::ConstantOp>(loc, intAttr));
+        continue;
+      }
+
+      return failure();
+    }
+
+    return success();
+  }
+
+  LogicalResult
+  matchAndRewrite(hw::AggregateConstantOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // Lower to concat.
+    SmallVector<Value> results;
+    if (failed(peelAttribute(op.getLoc(), adaptor.getFieldsAttr(), rewriter,
+                             results)))
+      return failure();
+    rewriter.replaceOpWithNewOp<comb::ConcatOp>(op, results);
+    return success();
+  }
+};
+
+struct HWArrayGetOpConversion : OpConversionPattern<hw::ArrayGetOp> {
+  using OpConversionPattern<hw::ArrayGetOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(hw::ArrayGetOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    SmallVector<Value> results;
+    auto arrayType = cast<hw::ArrayType>(op.getInput().getType());
+    auto elemType = arrayType.getElementType();
+    auto numElements = arrayType.getNumElements();
+    auto elemWidth = hw::getBitWidth(elemType);
+    if (elemWidth < 0)
+      return rewriter.notifyMatchFailure(op.getLoc(), "unknown element width");
+
+    auto lowered = rewriter.getRemappedValue(op.getInput());
```

Comment:
Why not use the adaptor? We wouldn't need the check below then.
```suggestion
    auto lowered = adaptor.getInput();
```

Reply:
Oh yeah. That works, thanks!